sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -79,16 +79,16 @@ case class Filter(condition: Expression, child: SparkPlan)
 
   // Split out all the IsNotNulls from condition.
   private val (notNullPreds, otherPreds) = splitConjunctivePredicates(condition).partition {
-    case IsNotNull(a) if child.output.contains(a) => true
+    case IsNotNull(a) if a.references.subsetOf(child.outputSet) => true
```

Comment:
This is not correct, for example, `IsNotNull(IsNull(a))` does not means `a` can not be `null`.